opte/src/engine/int_test.rs

```diff
@@ -99,6 +102,234 @@ fn next_block(offset: &[u8]) -> (&[u8], LegacyPcapBlock) {
+                "arp.rules_out" => $pav.port.num_rules("arp", Out),
+                "fw.flows_in" => $pav.port.num_flows("firewall", In),
+                "fw.flows_out" => $pav.port.num_flows("firewall", Out),
+                "fw.rules_in" => $pav.port.num_rules("firewall", In),
+                "fw.rules_out" => $pav.port.num_rules("firewall", Out),
+                "icmp.rules_in" => $pav.port.num_rules("icmp", In),
+                "icmp.rules_out" => $pav.port.num_rules("icmp", Out),
+                "nat.flows_in" => $pav.port.num_flows("nat", In),
+                "nat.flows_out" => $pav.port.num_flows("nat", Out),
+                "nat.rules_in" => $pav.port.num_rules("nat", In),
+                "nat.rules_out" => $pav.port.num_rules("nat", Out),
+                "router.rules_in" => $pav.port.num_rules("router", In),
+                "router.rules_out" => $pav.port.num_rules("router", Out),
+                "uft.flows_in" => $pav.port.num_flows("uft", In),
+                "uft.flows_out" => $pav.port.num_flows("uft", Out),
+                f => todo!("implement check for field: {}", f),
+            };
+            assert!(
+                *expected_val == actual_val,
+                "field value mismatch: field: {}, expected: {}, actual: {}",
+                field,
+                expected_val,
+                actual_val,
+            );
+        }
+
+        {
+            let expected = $pav.vps.epoch;
+            let actual = $pav.port.epoch();
+            assert!(
+                expected == actual,
+                "epoch mismatch: expected: {}, actual: {}",
+                expected,
+                actual,
+            );
+        }
+
+        {
+            let expected = $pav.vps.port_state;
+            let actual = $pav.port.state();
+            assert!(
+                expected == actual,
+                "port state mismatch: expected: {}, actual: {}",
+                expected,
+                actual,
+            );
+        }
+    };
+}
+
+macro_rules! incr_field {
+    ($vps:expr, $field:expr) => {
+        match $vps.counts.get_mut($field) {
+            Some(v) => *v += 1,
+            None => assert!(false, "field does not exist: {}", $field),
+        }
+    };
+}
+
+macro_rules! decr_field {
+    ($vps:expr, $field:expr) => {
+        match $vps.counts.get_mut($field) {
+            Some(v) => *v -= 1,
+            None => assert!(false, "field does not exist: {}", $field),
+        }
+    };
+}
+
+macro_rules! incr {
+    ($port_and_vps:expr, $fields:expr) => {
+        for f in $fields {
+            match f {
+                "epoch" => $port_and_vps.vps.epoch += 1,
+                _ => incr_field!($port_and_vps.vps, f),
+            }
+        }
+
+        assert_port!($port_and_vps);
+    };
+}
+
+// Like incr!, but do not assert_port! at the end. This is useful when
+// other state needs to be updated before asserting.
+macro_rules! incr_na {
```

Comment:
We could write `incr!` in terms of this macro, I think, if you wanted to trim a few lines.

Reply:
Good call.